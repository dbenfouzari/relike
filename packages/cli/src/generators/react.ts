import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
