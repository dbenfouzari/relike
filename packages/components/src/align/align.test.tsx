import { render } from "@testing-library/react";

import Alignment from "../alignment";
import Align from "./align";

type Direction = "flex-start" | "flex-end" | "center";

const expectDirections = (elm: HTMLElement, justifyContent: Direction | "", alignItems: Direction | "") => {
  const headerClass = elm.className;
  const headerRoot = document.getElementsByClassName(headerClass);
  const style = window.getComputedStyle(headerRoot[0]);

  expect(style.justifyContent).toBe(justifyContent);
  expect(style.alignItems).toBe(alignItems);
};

describe("Align", () => {
  it("should render successfully with top left", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.topLeft}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-start", "flex-start");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with top center", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.topCenter}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "center", "flex-start");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with top right", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.topRight}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-end", "flex-start");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with center left", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.centerLeft}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-start", "center");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with center", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.center}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "center", "center");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with center right", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.centerRight}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-end", "center");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with bottom left", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.bottomLeft}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-start", "flex-end");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with bottom center", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.bottomCenter}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "center", "flex-end");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with bottom right", () => {
    const { container, getByTestId } = render(
      <Align alignment={Alignment.bottomRight}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "flex-end", "flex-end");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render successfully with bad alignment", () => {
    const { container, getByTestId } = render(
      // @ts-expect-error
      <Align alignment={undefined}>
        <span>Hello Test</span>
      </Align>,
    );

    const divElement = getByTestId("align");

    expectDirections(divElement, "", "");
    expect(container.firstChild).toMatchSnapshot();
  });
});
