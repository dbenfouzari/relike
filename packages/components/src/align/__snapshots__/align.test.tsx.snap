// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Align should render successfully with bad alignment 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d9-0-2-11"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with bottom center 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d7-0-2-9"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with bottom left 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d6-0-2-8"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with bottom right 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d8-0-2-10"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with center 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d4-0-2-6"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with center left 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d3-0-2-5"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with center right 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d5-0-2-7"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with top center 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d1-0-2-3"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with top left 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d0-0-2-2"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;

exports[`Align should render successfully with top right 1`] = `
<div
  class="wrapper-0-2-1 wrapper-d2-0-2-4"
  data-testid="align"
>
  <span>
    Hello Test
  </span>
</div>
`;
